### PR TITLE
GO-4594: text gets deleted constantly when I'm typing

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
1. IsActive и IsFullDetailsSent returns slices now
2. IsFullDetailSent is not overwrite if it's already set for subscription
3. sFullDetailSent is set during subscriptions init (for sorted and simple sub)

### CODE REVIEW

File: util/maputils/map.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package maputils
```
[the remark this answers is not in the document]
Comment (AnastasiaShemyakinskaya) [reply util/maputils/map.go]: Fixed

File: core/subscription/cache.go [hunk 2]
```diff
@@ -17,62 +19,72 @@ type entry struct {
...
-		return false
-	}
-	for _, id := range subIds {
-		if pos := slice.FindPos(e.subIds, id); pos != -1 {
-			if !e.subIsActive[pos] {
-				return false
-			}
-		} else {
-			return false
+// GetActive indicates that entry is inside the current pagination window for all of provided subscription IDs
```
Comment (requilence) [inline core/subscription/cache.go:60]: comment is outdated

File: core/subscription/cache.go [hunk 3]
```diff
@@ -17,62 +19,72 @@ type entry struct {
...
-		return false
-	}
-	for _, id := range subIds {
-		if pos := slice.FindPos(e.subIds, id); pos != -1 {
-			if !e.subFullDetailsSent[pos] {
-				return false
-			}
-		} else {
-			return false
+// GetFullDetailsSent that in the context of ALL provided subscriptions we have previously sent the full ObjectSetDetails event
```
Comment (requilence) [inline core/subscription/cache.go:71]: comment is outdated

File: core/subscription/context.go [hunk 4]
```diff
@@ -202,6 +187,46 @@ func (ctx *opCtx) detailsEvents() {
...
+}, msgs []*pb.EventMessage) []*pb.EventMessage {
+	var sentAmendDetails, sentSetDetails []string
+	if prev != nil {
+		active := prev.GetActive()
+		detailsSent := prev.GetFullDetailsSent()
+		sentAmendDetails = lo.Intersect(active, detailsSent)
+
+		notActive := slice.Difference(info.subIds, active)
+		detailsNotSent := slice.Difference(info.subIds, detailsSent)
+		sentSetDetails = lo.Union(notActive, detailsNotSent)
```
Comment (requilence) [inline core/subscription/context.go:203]: naming("sent") is confusing here, it is actually opposite.  Maybe smth like "subIdsToSendFullDetails", "subIdsToSendDiffDetails"

File: core/subscription/context.go [hunk 5]
```diff
@@ -202,6 +187,46 @@ func (ctx *opCtx) detailsEvents() {
...
+		if len(sentSetDetails) != 0 {
+			msgs = ctx.addDetailsSentEvent(msgs, curr, sentSetDetails, info.keys)
+		}
+	} else {
+		msgs = ctx.addDetailsSentEvent(msgs, curr, info.subIds, info.keys)
+	}
+	return msgs
+}
+
+func (ctx *opCtx) addDetailsSentEvent(msgs []*pb.EventMessage, curr *entry, subIds, keys []string) []*pb.EventMessage {
```
Comment (requilence) [inline core/subscription/context.go:217]: incorrect method naming

File: core/subscription/cache.go [hunk 6]
```diff
@@ -17,62 +19,72 @@ type entry struct {
...
 		e.subIds = append(e.subIds, subId)
-		e.subIsActive = append(e.subIsActive, isActive)
-		e.subFullDetailsSent = append(e.subFullDetailsSent, isFullDetailSent)
+		e.subIsActive[subId] = isActive
+		e.subFullDetailsSent[subId] = isFullDetailSent
 	} else {
-		e.subIsActive[pos] = isActive
-		e.subFullDetailsSent[pos] = isFullDetailSent
+		e.subIsActive[subId] = isActive
+		if !e.subFullDetailsSent[subId] {
```
Comment (requilence) [inline core/subscription/cache.go:54]: lets add  an explaining comment here, why don't we overwrite

**Review by requilence — CHANGES_REQUESTED** [review]

**Review by requilence — APPROVED** [review]

**Pull request merged**